src/pcap/bosh_resolver.go

```diff
@@ -0,0 +1,208 @@
+package pcap
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"go.uber.org/zap"
+
+	"github.com/cloudfoundry/pcap-release/src/pcap/bosh"
+)
+
+type BoshResolver struct {
+	environment bosh.Environment
+	client      *http.Client
+	uaaURLs     []string
+	agentPort   int
+}
+
+func NewBoshResolver(environment bosh.Environment, agentPort int) (*BoshResolver, error) {
+	resolver := &BoshResolver{environment: environment, agentPort: agentPort} // TODO: get agent port from where?
+	err := resolver.setup()
+	if err != nil {
+		return nil, err
+	}
+	return resolver, err
+}
+
+func (br *BoshResolver) Name() string {
+	// TODO we need to differentiate between bosh resolvers for different environments (i.e. bootstrap-bosh, bosh. This would take some refactoring, e.g. in pcap.proto StatusResponse
+	// return fmt.Sprintf("bosh/%s", br.environment.Alias)
+	return "bosh"
+}
+
+func (br *BoshResolver) CanResolve(request *EndpointRequest) bool {
+	return request.GetBosh() != nil
+}
+
+func (br *BoshResolver) Resolve(request *EndpointRequest, log *zap.Logger) ([]AgentEndpoint, error) {
+	log = log.With(zap.String(LogKeyHandler, br.Name()))
+	log.Info("Resolving endpoints for bosh request")
+
+	err := br.validate(request)
+	if err != nil {
+		return nil, err
+	}
+
+	boshRequest := request.GetBosh()
+
+	err = br.authenticate(boshRequest.Token)
+	if err != nil {
+		return nil, err
+	}
+
+	instances, _, err := br.getInstances(boshRequest.Deployment, boshRequest.Token)
+	if err != nil {
+		log.Error("failed to get instances from bosh-director", zap.String(LogKeyTarget, ""))
+		return nil, err
+	}
+
+	var endpoints []AgentEndpoint
+	for _, instance := range instances {
+		identifier := strings.Join([]string{instance.Job, instance.Id}, "/")
+		endpoints = append(endpoints, AgentEndpoint{IP: instance.Ips[0], Port: br.agentPort, Identifier: identifier}) //TODO: defaultport ok here?
+	}
+
+	if len(endpoints) == 0 {
+		return nil, fmt.Errorf("no matching endpoints found")
+	}
+
+	log.Debug("received AgentEndpoints from Bosh Director", zap.Any("agent-endpoint", endpoints))
+	return endpoints, nil
+}
+
+func (br *BoshResolver) validate(endpointRequest *EndpointRequest) error {
```

Comment:
validate needs adaptation to the new semantics (e.g. the `Environment` needs to be set)